keps/sig-architecture/1194-prod-readiness/README.md

```diff
@@ -0,0 +1,166 @@
+  the merging of feature promotion PRs that lack production readiness.
+
+### Risks and Mitigations
+
+The primary risk is the slowing of feature merges. When this is due to the need
+for the developers to improve the quality of the feature, that is appropriate.
+When this is due to lack of bandwidth in the production readiness review team,
+that is harmful. To mitigate this, the implementation of this process must
+include a means of:
+ * Identifying potential community members for participation in the team
+ * A shadow program or other mechanism for preparing those individuals for
+   membership on the team
+ * Clear criteria for when one of these individuals is ready to become a full
+   participant
+ * Measurement of:
+   - Review throughput for production readiness reviews
+   - Team bench depth and size
+
+## Design Details
+
+### Phase 1 - Research and Pilot
+* Targeted to the 1.17 cycle.
+* Setup a pilot PRR team, which will:
+  * Deliver an initial PRR questionnaire and pilot it with non-blocking PRRs for
+    in-progress KEPs.
+  * Deliver an interview/questionnaire form for operators and interview them on
+    production issues that they experience, to ensure that the focus of this
+    effort is on meaningful problems.
+  * Deliver a postmortem summary of existing features that have stalled due to
+    production-readiness issues (e.g., cron jobs).
+* Resolve open questions, including:
+  * ~~Should the scope of this expand to feature lifecycle?~~ No, not at this
+    time.
+  * How do we measure the effectiveness of this effort?
+
+Status update on Phase 1:
+
+* Pilot was conducted in 1.17 and 1.18. The questionnaire has been through several
+  revisions and will be required in 1.19 as part of the KEP process.
+* Research was done via a survey, a report on which will be produced, presented
+  to SIG Architecture, and linked in here.
+* Follow-up interviews are in the planning stage as of the beginning of the 1.19
+  release cycle.
+* Post-mortem of past features has not been completed and is deferred until it
+  is deemed necessary by the subproject team.
+* Effectiveness measurement will come via a regular survey of users.
+
+### Phase 2 - Implementation
+
+The questionnaire was put into place in Phase 1, so what remains for Phase 2 is
+to document the review process and implement the tooling supporting it.
+
+The process will be enforced in the enhancements repository. When an enhancement
+is targeted at a release, a production readiness review will be necessary.
+Initially, the Enhancements Team will label the enhancement issue with a
+`prr-required` label; eventually tooling may do this automatically. The
+associated KEP must then be sure to add an approver from the
+`prod-readiness-review` team, which will be created in the `OWNERS_ALIASES` file
+of the enhancements repo. The enhancements verification CI must validate that a
+PRR approver is assigned to the KEP, and approves it prior to merging.
```

Comment:
What if the KEP isn't updated for Alpha->Beta or Beta->GA transition (e.g. all tests, graduation criteria, etc. was already filled in before?)

One solution that I can see is adding "target-stage: alpha|beta|ga" to kep.yaml and that would require a PR on every bump. Thoughts?

Reply:
#1620 adds `stage` and `latest-milestone`. We could alternatively add a milestone per stage: `alpha-milestone`, `beta-milestone`, `stable-milestone`. The issue would be that people would fill them out prior to starting work on those, and they could get stale.

Reply:
OK - I think with those changes to kep.yaml this bump is needed anyway, so I'm fine with that.

Reply:
@johnbelamaric -- I like that idea a lot. Maybe:

```
milestone:
- alpha: v1.14
- beta: v1.15
- stable: v1.18
```

so it's in the same data structure?

Reply:
👍

Reply:
+1 for Stephen`s suggestion

Reply:
OK. Instead of latest-milestone or in addition? suppose it goes beta in 1.19, with some change in 1.20, then stable in 1.21, what would the metadata look like?

Reply:
(Copying over from Slack)
@johnbelamaric -- yeah, I think `latest-milestone` is still good to have e.g., you can give a projection of timeline within `milestone` and the current status from `latest-milestone`.

Reply:
what about...

```
milestone:
- alpha: v1.16
- beta: v1.17
- stable: v1.20
- latest: v1.19
```

Or something like that (`active`?) to indicate the last update to the feature? This basically just rolls `latest-milestone` into there.

Reply:
Example:
```
latest-milestone: v1.17
milestone:
- alpha: v1.14
- beta: v1.15
- stable: v1.18
```

Reply:
Keeping `latest-milestone` separate I think makes it just a little easier to key on "Show me the KEPs for this milestone"

Reply:
So I think in totality I'd probably want to see:

```
latest-milestone: v1.17
stage: alpha
milestone:
- alpha: v1.14
- beta: v1.15
- stable: v1.18
```

`latest-milestone` and `stage` give you an enhancement's current status (assuming to that's up-to-date)

and then `milestone` allows you to query:
- when did something first enter a stage?
- when is this projected to graduate?

Reply:
works for me

Reply:
Closing the loop...
@johnbelamaric has added the metadata update as part of https://github.com/kubernetes/enhancements/pull/1620.

Reply:
#1620 was merged. Resolving.